test/cmocka/src/audio/pipeline/pipeline_free.c

```diff
@@ -0,0 +1,212 @@
+	list_item_append(&result.sched_comp->bsource_list,
+		&test_data->b1->sink_list);
+	test_data->b1->sink = result.sched_comp;
+	test_data->b1->source = test_data->second;
+
+	/*Testing component*/
+	pipeline_free(&result);
+
+	assert_ptr_equal(NULL, test_data->second->pipeline);
+	assert_ptr_equal(NULL, test_data->first->pipeline);
+}
+
+static void test_audio_pipeline_free_disconnect_upstream_list_delete
+	(void **state)
+{
+	struct pipeline_connect_data *test_data = *state;
+	struct pipeline result = test_data->p;
+	cleanup_test_data(test_data);
+
+	list_item_append(&result.sched_comp->bsource_list,
+		&test_data->b1->sink_list);
+	test_data->b1->sink = result.sched_comp;
+	test_data->b1->source = test_data->second;
+
+	/*Testing component*/
+	pipeline_free(&result);
+
+	assert_true(list_is_empty(&test_data->second->bsource_list));
+	assert_true(list_is_empty(&test_data->first->bsource_list));
+}
+
+static void test_audio_pipeline_free_disconnect_downstream_full(void **state)
+{
+	struct pipeline_connect_data *test_data = *state;
+	struct pipeline result = test_data->p;
+	cleanup_test_data(test_data);
+
+	/*Set pipeline to check that is null later*/
+	test_data->first->pipeline = &result;
+	test_data->second->pipeline = &result;
+	test_data->second->comp.pipeline_id = PIPELINE_ID_SAME;
+	test_data->first->comp.pipeline_id = PIPELINE_ID_SAME;
+	test_data->b1->source = test_data->first;
+	list_item_append(&result.sched_comp->bsink_list,
+		&test_data->b1->source_list);
+	test_data->b1->sink = test_data->second;
+
+	/*Testing component*/
+	pipeline_free(&result);
+
+	assert_ptr_equal(NULL, test_data->second->pipeline);
+	assert_ptr_equal(NULL, test_data->first->pipeline);
+}
+
+static void test_audio_pipeline_free_disconnect_downstream_list_del
+(void **state)
+{
+	struct pipeline_connect_data *test_data = *state;
+	struct pipeline result = test_data->p;
+	cleanup_test_data(test_data);
+
+	test_data->b1->source = test_data->first;
+	list_item_append(&result.sched_comp->bsink_list,
+		&test_data->b1->source_list);
+	test_data->b1->sink = test_data->second;
+
+	/*Testing component*/
+	pipeline_free(&result);
+
+	assert_true(list_is_empty(&test_data->second->bsink_list));
+	assert_true(list_is_empty(&test_data->first->bsink_list));
+}
+
+
+
+int main(void)
+{
+
+	const struct CMUnitTest tests[] = {
+		cmocka_unit_test(
+			test_audio_pipeline_free_comp_busy
+		),
+		cmocka_unit_test(
+			test_audio_pipeline_free_return_value
+		),
+		cmocka_unit_test(
+			test_audio_pipeline_free_sheduler_task_free
+		),
+		cmocka_unit_test(
+			test_audio_pipeline_free_disconnect_upstream_full
+		),
+		cmocka_unit_test(
+			test_audio_pipeline_free_disconnect_upstream_list_delete
+		),
+		cmocka_unit_test(
+			test_audio_pipeline_free_disconnect_downstream_full
+		),
+		cmocka_unit_test(
+			test_audio_pipeline_free_disconnect_downstream_list_del
+		),
+	};
+
+
+	cmocka_set_message_output(CM_OUTPUT_TAP);
+
+	return cmocka_run_group_tests(tests, setup, teardown);
```

Comment:
You use setup and teardown for group that is not linked at all - it looks like you reset state in every test.
To show intent and 'scope' of state it would be better to use cmocka_unit_test_setup_teardown in the future for cases like this. Group setup is better for flow when test2 depends on test1 etc.